File: Sources/termfarm/Persistence.swift
import Foundation

func saveURL() -> URL {
    let home = FileManager.default.homeDirectoryForCurrentUser
    let dir  = home.appendingPathComponent(".local/share/termfarm", isDirectory: true)

    try? FileManager.default.createDirectory(
        at: dir,
        withIntermediateDirectories: true
    )

    return dir.appendingPathComponent("save.json")
}

func loadFarm() throws -> FarmState {
    let url = saveURL()
        let data = try Data(contentsOf: url)

        return try JSONDecoder().decode(FarmState.self, from: data)
}

func saveFarm(_ farm: FarmState) throws {
    let data = try JSONEncoder().encode(farm)
    try data.write(to: saveURL(), options: .atomic)
}

File: internal/persistence.go
package internal

import (
	"encoding/json"
	"os"
	"path/filepath"
)

func saveURL() string {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".local", "share", "termfarm")

	os.MkdirAll(dir, 0755)

	return filepath.Join(dir, "save.json")
}

func loadFarm() (FarmState, error) {
	url := saveURL()
	data, err := os.ReadFile(url)
	if err != nil {
		return FarmState{}, err
	}

	var farm FarmState
	if err := json.Unmarshal(data, &farm); err != nil {
		return FarmState{}, err
	}

	updateMarketIfNeeded(&farm)

	return farm, nil
}
